src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1870,16 +1870,30 @@ def vnc_bind_choices(self):
     Method called when system is ready, supposed to start VMs
     flagged that way.
     """
-    if args['id'] != 'ready':
-        return
-
-    global ZFS_ARC_MAX_INITIAL
-    ZFS_ARC_MAX_INITIAL = sysctl.filter('vfs.zfs.arc.max')[0].value
-
-    await middleware.call('vm.initialize_vms')
-
-    for vm in await middleware.call('vm.query', [('autostart', '=', True)]):
-        await middleware.call('vm.start', vm['id'])
+    if args['id'] == 'ready':
+        global ZFS_ARC_MAX_INITIAL
+        ZFS_ARC_MAX_INITIAL = sysctl.filter('vfs.zfs.arc.max')[0].value
+
+        await middleware.call('vm.initialize_vms')
+
+        for vm in await middleware.call('vm.query', [('autostart', '=', True)]):
+            await middleware.call('vm.start', vm['id'])
+    elif args['id'] == 'shutdown':
+        for vm in await middleware.call('vm.query', [('state', '=', 'RUNNING')]):
+            stop_job = await middleware.call('vm.stop', vm['id'])
```

Comment:
We should run stop concurrently or it can add up to several minutes if there are many of them.